carma-messenger-core/j2735_convertor/src/psm_convertor.cpp

```diff
@@ -0,0 +1,271 @@
+{
+  // Convert semi_major Axis
+  out_msg.semi_major = ValueConvertor::valueCavToJ2735<uint8_t>(
+      in_msg.semi_major, units::TWENTIETH_M_PER_M, in_msg.presence_vector,
+      carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_AVAILABLE, j2735_v2x_msgs::msg::PositionalAccuracy::ACCURACY_UNAVAILABLE);
+
+  // Convert semi_minor Axis
+  out_msg.semi_minor = ValueConvertor::valueCavToJ2735<uint8_t>(
+      in_msg.semi_minor, units::TWENTIETH_M_PER_M, in_msg.presence_vector,
+      carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_AVAILABLE, j2735_v2x_msgs::msg::PositionalAccuracy::ACCURACY_UNAVAILABLE);
+
+  // Convert Orientation
+  out_msg.orientation = ValueConvertor::valueCavToJ2735<uint16_t>(
+      in_msg.orientation, units::DEG_360_OVER_65535_PER_DEG, in_msg.presence_vector,
+      carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_ORIENTATION_AVAILABLE,
+      j2735_v2x_msgs::msg::PositionalAccuracy::ACCURACY_ORIENTATION_UNAVAILABLE);
+}
+
+void PSMConvertor::convert(const carma_v2x_msgs::msg::PathHistory& in_msg, j2735_v2x_msgs::msg::PathHistory& out_msg)
+{
+    out_msg.curr_gnss_status = in_msg.curr_gnss_status;
+    out_msg.initial_position.lat.latitude = in_msg.initial_position.lat.latitude / latitude_conversion_const_;
+    out_msg.initial_position.lon.longitude = in_msg.initial_position.lon.longitude / longitude_conversion_const_;
+    out_msg.initial_position.elevation.elevation = in_msg.initial_position.elevation.elevation / elevation_conversion_const_;
+    out_msg.initial_position.utc_time = in_msg.initial_position.utc_time;
+    out_msg.initial_position.time_confidence = in_msg.initial_position.time_confidence;
+    out_msg.initial_position.pos_confidence = in_msg.initial_position.pos_confidence;
+    out_msg.initial_position.speed_confidence = in_msg.initial_position.speed_confidence;
+
+    convert(in_msg.initial_position.heading, out_msg.initial_position.heading);
+    convert(in_msg.initial_position.speed.speed, out_msg.initial_position.speed.speed);
+    convert(in_msg.initial_position.pos_accuracy, out_msg.initial_position.pos_accuracy);
+
+    for (auto i : out_msg.crumb_data.points) {
+
+        j2735_v2x_msgs::msg::PathHistoryPoint j;
+
+        j.lat_offset.offset = i.lat_offset.offset / 1e-7;
+        j.lon_offset.offset = i.lon_offset.offset / 1e-7;
+        j.elevation_offset.offset = i.elevation_offset.offset / 0.1;
+        j.time_offset.offset = i.time_offset.offset / 0.01;
+        j.heading.heading = i.heading.heading / heading_conversion_const_;
+        j.speed.speed = i.speed.speed / 0.02;
+        j.pos_accuracy.semi_major = i.pos_accuracy.semi_major / 0.05;
+        j.pos_accuracy.orientation = i.pos_accuracy.orientation / 0.0054932479;
+        out_msg.crumb_data.points.push_back(j);
+    }
+}
+
+void PSMConvertor::convert(const carma_v2x_msgs::msg::PSM& in_msg, j2735_v2x_msgs::msg::PSM& out_msg)
+{
+      convert(in_msg.accel_set, out_msg.accel_set);
```

Comment:
Same note as initial_position here. Some of these fields are optional and have presence_vectors that need to be checked and assigned